Analysis/Tasks/PWGHF/DplusDminusCorrelator.cxx

```diff
@@ -0,0 +1,876 @@
+  {
+    registry.add("hcountDplustriggersMCGen", "Dplus trigger particles - MC gen;;N of trigger D0", {HistType::kTH2F, {{1, -0.5, 0.5}, {(std::vector<double>)bins, "#it{p}_{T} (GeV/#it{c})"}}});
+  }
+
+  void process(aod::McCollision const& mccollision, soa::Join<aod::McParticles, aod::HfCandProng3MCGen> const& particlesMC)
+  {
+    int counterDplusDminus = 0;
+    registry.fill(HIST("hMCEvtCount"), 0);
+    //MC gen level
+    for (auto& particle1 : particlesMC) {
+      //check if the particle is Dplus or Dminus (both can be trigger) - NOTE: decay channel is not probed!
+      if (std::abs(particle1.pdgCode()) != 411) {
+        continue;
+      }
+      if (cutYCandMax >= 0. && std::abs(RecoDecay::Y(array{particle1.px(), particle1.py(), particle1.pz()}, RecoDecay::getMassPDG(particle1.pdgCode()))) > cutYCandMax) {
+        continue;
+      }
+      if (cutPtCandMin >= 0. && std::abs(particle1.pt()) < cutPtCandMin) {
+        continue;
+      }
+
+      registry.fill(HIST("hPtCandMCGen"), particle1.pt());
+      registry.fill(HIST("hEtaMCGen"), particle1.eta());
+      registry.fill(HIST("hPhiMCGen"), particle1.phi());
+      registry.fill(HIST("hYMCGen"), RecoDecay::Y(array{particle1.px(), particle1.py(), particle1.pz()}, RecoDecay::getMassPDG(particle1.pdgCode())));
+      counterDplusDminus++;
+
+      //D-Dbar correlation dedicated section
+      //if it's Dplus/Dminus, search for LS pair companions and evaluate correlations.
+      registry.fill(HIST("hcountDplustriggersMCGen"), 0, particle1.pt()); //to count trigger Dplus (normalisation)
+      for (auto& particle2 : particlesMC) {
+        if (std::abs(particle2.pdgCode()) != 411) { //check that associated is a Dplus/Dminus (both are fine)
+          continue;
+        }
+        if (cutYCandMax >= 0. && std::abs(RecoDecay::Y(array{particle2.px(), particle2.py(), particle2.pz()}, RecoDecay::getMassPDG(particle2.pdgCode()))) > cutYCandMax) {
+          continue;
+        }
+        if (cutPtCandMin >= 0. && std::abs(particle2.pt()) < cutPtCandMin) {
+          continue;
+        }
+        if (particle2.pdgCode() == particle1.pdgCode()) { //like-sign condition (both 411 or both -411) and pT_Trig>pT_assoc
+          //Excluding self-correlations (in principle not possible due to the '<' condition, but could rounding break it?)
+          if (particle1.mRowIndex == particle2.mRowIndex) {
+            continue;
+          }
+          entryDplusDminusPair(getDeltaPhi(particle2.phi(), particle1.phi()),
+                               particle2.eta() - particle1.eta(),
+                               particle1.pt(),
+                               particle2.pt());
+        }
+      } // end inner loop
+    }   //end outer loop
+    registry.fill(HIST("hcountDplusDminusPerEvent"), counterDplusDminus);
+  }
+};
+
+/// c-cbar correlator table builder - for MC gen-level analysis
+struct CCbarCorrelatorMCGen {
+
+  Produces<aod::DDbarPair> entryccbarPair;
```

Comment:
Why `ccbar`?

Reply:
Not sure I got the comment, here I'm doing c-cbar parton-level correlations, which can safely exploit the aod::DDbarPair table (the same columns are needed), so I'm using that table without creating a new one in HFSecondaryVertex.h with exactly the same definition

Reply:
Hmm, ok, that makes me wonder. How to you intend to handle the multiple instances of `DDbarPair` produced by different tasks?

Reply:
Indeed, I have to run a single task at each time to avoid conflicts (e.g. either data-like, or MC gen, or ccbar, etc)